### PR TITLE
12-Headers Class

### PR DESCRIPTION
**Pull request by KyoshiNoda**
## Description
This pull request introduces the `Headers` class, which represents a collection of `Header` objects.

- The `Headers` class uses a `HashMap` of type: `HashMap<String, ArrayList<Header>>`. 
  - Each key in the map represents a unique identifier.
  - The associated value is a list of `Header` objects, all sharing the same key but potentially having different values.
- The `ArrayList<Header>` ensures that multiple headers with the same key can be stored and maintained in insertion order.
- This implementation supports adding, removing, and retrieving headers based on their keys.

### Key Features:
- Each key maps to a collection (`ArrayList`) of `Header` objects.
- All `Header` objects in a given list share the same key but can have different values.

### Screenshot
Screenshot 2024-12-27 at 9 15 22 PM

Screenshot 2024-12-27 at 9 15 08 PM

### Story Reference:
- Story #12
- Docs: [Apache Kafka Headers](https://kafka.apache.org/25/javadoc/org/apache/kafka/common/header/Headers.html)

### CODE REVIEW

File: src/main/java/commons/headers/Headers.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        } else {
+            hashMap.get(key).add(header);
+        }
+        return this;
+    }
+
+    @Override
+    public HeadersInterface remove(String key) throws IllegalStateException {
+        // Key does not exist
+        if (!hashMap.containsKey(key)) {
```
Comment (J-Obog) [inline src/main/java/commons/headers/Headers.java:39]: should this throw an error if the key isn't found?

File: src/main/java/commons/headers/Headers.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    @Override
+    public HeadersInterface add(Header header) throws IllegalStateException {
+        String key = header.getKey();
+        // First time keys
+        if (!hashMap.containsKey(key)) {
+            hashMap.put(key, new ArrayList<>());
+            hashMap.get(key).add(header);
+        } else {
+            hashMap.get(key).add(header);
+        }
```
Comment (J-Obog) [inline src/main/java/commons/headers/Headers.java:32]: Very minor nit: You can do

```
if (!hashMap.containsKey(key)) {
            hashMap.put(key, new ArrayList<>());
        } 

hashMap.get(key).add(header);
```

File: src/main/java/commons/headers/Headers.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            throw new IllegalArgumentException("Key not found.");
+        }
+        return hashMap.get(key);
+    }
+
+    @Override
+    public Header lastHeader(String key) {
+        // Key does not exist
+        if (!hashMap.containsKey(key)) {
+            throw new IllegalArgumentException("Key not found.");
```
Comment (J-Obog) [inline src/main/java/commons/headers/Headers.java:60]: similar question as https://github.com/tferdous17/flux/pull/29/files#r1898777566

File: src/main/java/commons/headers/Headers.java (hunk 1 above)
Comment (KyoshiNoda) [reply src/main/java/commons/headers/Headers.java:39]: I think we should, @tferdous17 what's your input?

File: src/main/java/commons/headers/Headers.java (hunk 3 above)
Comment (KyoshiNoda) [reply src/main/java/commons/headers/Headers.java:60]: I think we should, @tferdous17 what's your input?

File: src/main/java/commons/headers/Headers.java (hunk 1 above)
Comment (tferdous17) [reply src/main/java/commons/headers/Headers.java:39]: yeah i think throwing an error for now is fine as default behavior if a key DNE

File: src/main/java/commons/headers/Headers.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
+package commons.headers;
+
+import commons.header.Header;
+
+import java.util.*;
+
+public class Headers implements HeadersInterface {
+    private final HashMap<String, ArrayList<Header>> hashMap = new LinkedHashMap<>();
```
Comment (tferdous17) [inline src/main/java/commons/headers/Headers.java:8]: hmm not 100% certain if we should have the values be `ArrayList<Header>`
I was thinking each key-value pair would be at most 1 key, 1 value, lmk your thought process on the decision though

Comment (KyoshiNoda) [reply src/main/java/commons/headers/Headers.java:8]: My thought process was.... For each `key` there will be a list of Header that all share the same key with different values... Take the following example:

```javascript
Key: "Accept"
Value: [
    Header(key="Accept", value="text/html"),
    Header(key="Accept", value="application/json"),
    Header(key="Accept", value="image/png")
]

Key: "Set-Cookie"
Value: [
    Header(key="Set-Cookie", value="sessionId=abc123"),
    Header(key="Set-Cookie", value="userId=kyoshi"),
    Header(key="Set-Cookie", value="theme=dark")
]

Key: "Authorization"
Value: [
    Header(key="Authorization", value="Bearer token123")
]

Key: "Content-Type"
Value: [
    Header(key="Content-Type", value="application/json")
]
```

Comment (tferdous17) [reply src/main/java/commons/headers/Headers.java:8]: ok yeah I get your justification, lgtm then

**Review by tferdous17 — APPROVED** [review]
lgtm

**Review by J-Obog — APPROVED** [review]
LGTM

**Pull request merged**